lib/rspec/core/hooks.rb

```diff
@@ -503,6 +503,8 @@ def all_hooks_for(position, scope)
         end
 
         def run_owned_hooks_for(position, scope, example_or_group)
+          return if example_or_group.class.respond_to?(:metadata) &&
+                      example_or_group.class.metadata.keys.include?(:skip)
```

Comment:
There's a few things about this that are a bit sub-optimal:

1. `metadata.keys.include?(:skip)` is a linear operation, scanning an array of metadata keys.  It's more efficient to use `Hash#key?` by doing `metadata.key?(:skip)` as that's a constant time check.
2. Checking `respond_to?` is sub-optimal, but I see why it's necessary here, because this is a central place called by both example groups and examples.  Instead, I think you can move this check to the call site of this method for the `:context` hook case.  Here's a diff that does that:

``` diff
diff --git a/lib/rspec/core/hooks.rb b/lib/rspec/core/hooks.rb
index 4a69b76..0f77064 100644
--- a/lib/rspec/core/hooks.rb
+++ b/lib/rspec/core/hooks.rb
@@ -456,7 +456,9 @@ module RSpec
           return if RSpec.configuration.dry_run?
 
           if scope == :context
-            run_owned_hooks_for(position, :context, example_or_group)
+            unless example_or_group.class.metadata.key?(:skip)
+              run_owned_hooks_for(position, :context, example_or_group)
+            end
           else
             case position
             when :before then run_example_hooks_for(example_or_group, :before, :reverse_each)
@@ -503,8 +505,6 @@ module RSpec
         end
 
         def run_owned_hooks_for(position, scope, example_or_group)
-          return if example_or_group.class.respond_to?(:metadata) &&
-                      example_or_group.class.metadata.keys.include?(:skip)
           matching_hooks_for(position, scope, example_or_group).each do |hook|
             hook.run(example_or_group)
           end
```

Notice that there's no need to check `respond_to?` in this case.